Source/WebKit/UIProcess/API/gtk/DropTargetGtk4.cpp

```diff
@@ -129,20 +131,55 @@ void DropTarget::accept(GdkDrop* drop, std::optional<WebCore::IntPoint> position
         }, this);
     }
 
+    static const char* const portalMIMETypes[] = {
+        "application/vnd.portal.filetransfer",
+        "application/vnd.portal.files", // Deprecated, but added for compatibility
+    };
+
     static const char* const supportedMimeTypes[] = {
+        "application/vnd.portal.filetransfer",
+        "application/vnd.portal.files", // Deprecated, but added for compatibility
         "text/html",
         "_NETSCAPE_URL",
         "text/uri-list",
         "application/vnd.webkitgtk.smartpaste",
         "org.webkitgtk.WebKit.custom-pasteboard-data"
     };
 
+    bool transferredFilesFromPortal = false;
     for (unsigned i = 0; i < G_N_ELEMENTS(supportedMimeTypes); ++i) {
         if (!gdk_content_formats_contain_mime_type(formats, supportedMimeTypes[i]))
             continue;
 
+        // Reading from the File Transfer portal is a bit special. When either portal
+        // mimetypes are present, GTK serializes them using the GdkFileList type. If
+        // this type is present, ignore file:// URIs from the "text/uri-list" later on.
+        if (!transferredFilesFromPortal && g_strv_contains(portalMIMETypes, supportedMimeTypes[i])) {
+            ASSERT(gdk_content_formats_contain_gtype(formats, GDK_TYPE_FILE_LIST));
+
+            m_dataRequestCount++;
+            loadData([this, cancellable = m_cancellable](Vector<GRefPtr<GFile>>&& files) {
+                if (g_cancellable_is_cancelled(cancellable.get()))
+                    return;
+
+                // Convert files transferred by the File Transfer portal into URIs
+                if (!files.isEmpty()) {
+                    for (auto& file : files) {
+                        GUniquePtr<char> uri(g_file_get_uri(file.get()));
+                        if (!m_uriListBuilder.isEmpty())
+                            m_uriListBuilder.append("\r\n");
+                        m_uriListBuilder.append(String::fromUTF8(uri.get()));
```

Comment:
So, the files aren't really needed, we just need the uri, in that case I think it's easier if the completion handler receives a `Vector<String>&&`